### PR TITLE
Add Grid Types and Arrangement Transformations

### PR DESCRIPTION
**Pull request by cl445**
## Summary
- Introduces layout/placement types that separate key behavior from key position (Phase 1.9–1.12)
- `GridArrangement` transformations enable deriving utility-left/landscape layouts from a single base definition
- Depends on PR #144 (Core Value Types)

## New Types
| Type | Purpose |
|------|---------|
| `KeyPlacement` | References a key by ID with width/height multipliers |
| `GridArrangement` | Spatial arrangement of keys + `mirroredHorizontally()`, `resized()`, `removing()` |
| `ArrangementContext` | Portrait/landscape × utility left/right (4 contexts) |

## Test plan
- [x] 20 new unit tests in `GridTypeTests.swift`
- [x] `KeyPlacement`: default multipliers, Codable, Equatable
- [x] `mirroredHorizontally()`: row reversal, multiplier preservation, double-mirror = identity
- [x] `resized()`: target key adjusted, others untouched, heightMultiplier preserved
- [x] `removing()`: key removed from all rows, no-op for missing key, columns unchanged
- [x] Edge cases: empty grid, single-cell grid
- [x] Full test suite passes (375 tests, 0 failures)

## New Features
* Introduced configurable keyboard arrangement contexts supporting portrait and landscape orientations with utility panel placement options
* Added grid-based layout system enabling key placement configuration with resizable and repositionable keys

## Tests
* Added comprehensive test coverage for arrangement contexts, grid layouts, and key placement configurations

### CODE REVIEW

File: wurstfingerKeyboard/GridArrangement.swift [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import Foundation
+
+/// A concrete spatial arrangement of keys.
+/// Multiple arrangements can lay out the same key pool differently.
+struct GridArrangement: Codable, Equatable {
+    /// Number of logical columns.
+    /// The sum of widthMultipliers in each row (plus columns spanned by keys from
+    /// previous rows via heightMultiplier) must equal this value.
+    let columns: Int
```
[the remark this answers is not in the document]
Comment (cl445) [reply wurstfingerKeyboard/GridArrangement.swift:16]: Addressed in commit b90cf7f:
- Added `precondition(newColumns > 0)` and `precondition(newWidth > 0)` to `resized()`
- Added `precondition(widthMultiplier > 0)` and `precondition(heightMultiplier > 0)` to `KeyPlacement.init`

Intentionally omitted the `keyExists` precondition: `resized()` is a pure transformation, and if the key doesn't exist the result is simply unchanged rows with the new column count. The validation layer (`ValidationError.columnMismatch`) already catches any resulting inconsistencies, keeping the transformation composable.

File: wurstfingerKeyboard/KeyPlacement.swift [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    /// Enables e.g. a tall return key spanning 2 rows.
+    let heightMultiplier: Int
+
+    init(keyId: String, widthMultiplier: Int = 1, heightMultiplier: Int = 1) {
+        precondition(widthMultiplier > 0, "widthMultiplier must be positive")
+        precondition(heightMultiplier > 0, "heightMultiplier must be positive")
+        self.keyId = keyId
+        self.widthMultiplier = widthMultiplier
+        self.heightMultiplier = heightMultiplier
+    }
```
[the remark this answers is not in the document]
Comment (cl445) [reply wurstfingerKeyboard/KeyPlacement.swift:29]: Good catch! Added a custom `init(from:)` that validates multipliers and throws `DecodingError.dataCorruptedError` for non-positive values, plus two tests for zero/negative multiplier decoding. Addressed in commit b0b7667.

File: wurstfingerTests/CoreValueTypeTests.swift [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+            #expect(decoded == slide)
+        }
+    }
+
+    @Test func keyStyleAllCases() throws {
+        for style in [KeyStyle.primary, .secondary, .utility, .spacebar, .accent] {
+            let data = try JSONEncoder().encode(style)
+            let decoded = try JSONDecoder().decode(KeyStyle.self, from: data)
+            #expect(decoded == style)
+        }
```
[the remark this answers is not in the document]
Comment (cl445) [reply wurstfingerTests/CoreValueTypeTests.swift:406]: Declining this one — the explicit encode/decode/assert pattern keeps each test self-contained and easy to read at a glance. A shared helper adds indirection without meaningful reduction in test count.
